protocol/bc/vmcontext.go

```diff
@@ -0,0 +1,89 @@
+package bc
+
+import (
+	"bytes"
+
+	"chain/protocol/vm"
+)
+
+func NewBlockVMContext(block *Block, prog []byte, args [][]byte) *vm.Context {
+	blockHash := block.Hash().Bytes()
+	return &vm.Context{
+		VMVersion: 1,
+		Code:      prog,
+		Arguments: args,
+
+		BlockHash:            &blockHash,
+		BlockTimeMS:          &block.TimestampMS,
+		NextConsensusProgram: &block.ConsensusProgram,
+	}
+}
+
+func NewTxVMContext(tx *Tx, index uint32, prog Program, args [][]byte) *vm.Context {
+	var (
+		txSigHash        = tx.SigHash(index).Bytes()
+		numResults       = uint64(len(tx.Results))
+		assetID          = tx.Inputs[index].AssetID()
+		assetIDBytes     = assetID[:]
+		amount           = tx.Inputs[index].Amount()
+		inputRefDataHash = hashData(tx.Inputs[index].ReferenceData).Bytes()
+		txRefDataHash    = hashData(tx.ReferenceData).Bytes()
+	)
+
+	checkOutput := func(index uint64, refdatahash []byte, amount uint64, assetID []byte, vmVersion uint64, code []byte) (bool, error) {
+		if index >= uint64(len(tx.Outputs)) {
+			return false, vm.ErrBadValue
```

Comment:
The spec says

> Fails if the number of outputs is less or equal to index.

which I interpret to explicitly require the current behavior.